src/server/src/main/java/io/cassandrareaper/service/Heart.java

```diff
@@ -0,0 +1,188 @@
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import javax.management.JMException;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+final class Heart implements AutoCloseable {
+
+  private static final AtomicBoolean GUAGES_REGISTERED = new AtomicBoolean(false);
+  private static final Logger LOG = LoggerFactory.getLogger(Heart.class);
+  private static final long DEFAULT_MAX_FREQUENCY = TimeUnit.MINUTES.toMillis(1);
+
+  private final AtomicLong lastBeat = new AtomicLong(System.currentTimeMillis() - TimeUnit.HOURS.toMillis(1));
+  private final ForkJoinPool forkJoinPool = new ForkJoinPool(64);
+  private final long maxBeatFrequencyMillis;
+  private final AtomicBoolean updatingNodeMetrics = new AtomicBoolean(false);
+
+  private Heart(long maxBeatFrequency) {
+    this.maxBeatFrequencyMillis = maxBeatFrequency;
+  }
+
+  static Heart create() {
+    return new Heart(DEFAULT_MAX_FREQUENCY);
+  }
+
+  @VisibleForTesting
+  static Heart create(long maxBeatFrequencyMillis) {
+    return new Heart(maxBeatFrequencyMillis);
+  }
+
+  synchronized void beat(AppContext context) {
+    if (context.storage instanceof IDistributedStorage
+        && lastBeat.get() + maxBeatFrequencyMillis < System.currentTimeMillis()) {
+
+      lastBeat.set(System.currentTimeMillis());
+      ((IDistributedStorage) context.storage).saveHeartbeat();
+
+      if (ReaperApplicationConfiguration.DatacenterAvailability.EACH == context.config.getDatacenterAvailability()) {
+        updateAllReachableEmptyNodeMetrics(context);
+      }
+    }
+  }
+
+  AtomicBoolean isCurrentlyUpdatingNodeMetrics() {
+    return new AtomicBoolean(updatingNodeMetrics.get());
+  }
+
+  @Override
+  public void close() {
+    try {
+      forkJoinPool.shutdown();
+      forkJoinPool.awaitTermination(10, TimeUnit.SECONDS);
+    } catch (InterruptedException ignore) {
+    } finally {
+      forkJoinPool.shutdownNow();
+    }
+  }
+
+  private void updateAllReachableEmptyNodeMetrics(AppContext context) {
+    Preconditions.checkArgument(context.storage instanceof IDistributedStorage);
+    IDistributedStorage storage = ((IDistributedStorage) context.storage);
+    registerGuages(context);
+
+    if (!updatingNodeMetrics.getAndSet(true)) {
+      HostConnectionCounters hostConnectionCounters = context.jmxConnectionFactory.getHostConnectionCounters();
+      int jmxTimeoutSeconds = context.config.getJmxConnectionTimeoutInSeconds();
+
+      forkJoinPool.submit(() -> {
+        try (Timer.Context t0 = timer(context, "metrics", "all")) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "metrics" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)